Implementation/59970 introduce enterprise banner viewcomponent

ulferts opened the pull request:
# Ticket

https://community.openproject.org/wp/59970

# What are you trying to accomplish?

Introduce a ViewComponent to be used when rendering a banner. It will be used first within the project lifecycle administration but should ideally be used throughout the application eventually.

## Screenshots

<img width="783" alt="image" src="https://github.com/user-attachments/assets/be39a098-c62c-4350-92a7-c517a3e60b30">


# What approach did you choose and why?

The component enforces i18n and link href to be at a standard location so that it is easier to find the appropriate place for this kind of information.

# Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

ulferts committed:
remove now duplicate i18n key

HDinger commented on app/components/enterprise_edition/banner_component.html.erb:
```diff
@@ -0,0 +1,19 @@
+<%=
+  render(Primer::OpenProject::GridLayout.new(css_class: "op-ee-banner", **@system_arguments)) do |grid|
```
You can use the `grid_layout` helper for this.

ulferts replied:
Thanks for the pointer

HDinger commented on app/components/tables/column_header_component.rb:
```diff
@@ -0,0 +1,131 @@
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+class Tables::ColumnHeaderComponent < ApplicationComponent
```
Should that file be part of the PR? 🤔

ulferts replied:
Sorry for that. You are right of course, that this found its way in here by mistake.

HDinger commented on app/components/enterprise_edition/banner_component.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+
+    def description
+      @description || I18n.t("ee.upsale.#{feature_key}.description")
```
There should be a default for the description I guess, similar to how title and link_title have one.

ulferts replied:
I believe that while title and link text will more or less always stay the same, the description needs to be different for every feature we apply it to. 

To manifest that believe, the component now raises an error with a pointer on how to fix this.

HDinger commented on app/components/enterprise_edition/banner_component.sass:
```diff
@@ -0,0 +1,65 @@
+  &--icon
+    width: 32px
+    height: 32px
+    margin-top: -6px
+    margin-left: -5px
```
This is somehow bad style, as these numbers are completly random and will break as soon as we change the layout of the shield in any way. You could get rid of that by using a flexBox on the shield with `align_items: center` and `justify_content: center`

ulferts replied:
Done, thanks.

HDinger commented on lookbook/previews/open_project/enterprise_edition/banner_component_preview.rb:
```diff
@@ -0,0 +1,79 @@
+      def default
+        render(
+          ::EnterpriseEdition::BannerComponent
+            .new(:form_configuration,
```
`form_configuration` is maybe a bad example key, as the content expects a list afterwards and it looks to have a colon but nothing coming afterwards.
<img width="522" alt="Bildschirmfoto 2024-12-11 um 13 14 24" src="https://github.com/user-attachments/assets/45490e0f-79dd-47cb-aa5e-aefc2e4ba88f">

ulferts replied:
Yes, that looks bad. I wanted to replace it with a different one. But the one that was available didn't have an href set so I failed first. Now I find out that this alternative is also outdated so I removed it in the PR.

Ideally, it will be replaced by the first feature that actually uses this component, @toy. So when the component is applied to the project lifecycle administration, we can use that as the example.

HDinger commented on app/components/enterprise_edition/banner_component.sass:
```diff
@@ -0,0 +1,65 @@
+
+  &--shield
+    @extend .upsale-border-colored
+    width: 32px
```
Since that width is used three times in that file (and it is important that it is always the same value), I'd like to extract that into a variable

ulferts replied:
:+1:

HDinger reviewed (changes requested):
I like the idea of using a key and thus enforcing the same structure of localisation for each feature. This should also make it easier to use. 👍 
I am wondering however, whether we should allow it at all to overwrite the key-based strings, but probably there are use cases where that makes sense.. 🤷‍♀️ 

One minor thing: It looks like the text is slighly not centred compared to the shield. This is due to the shield being smaller than the grid next to it. 
<img width="561" alt="Bildschirmfoto 2024-12-11 um 13 39 47" src="https://github.com/user-attachments/assets/e51b1405-9cd0-4c7b-a829-0c79425cf1e5">

ulferts committed:
add enterprise edition banner component

ulferts committed:
remove outdated ee i18n

ulferts commented:
I believe to have addressed all your remarks, @HDinger .

Since the PR included an erroneously added file and the PR in total isn't big, I force pushed. I hope that does not make the rereview too strenuous.

> I am wondering however, whether we should allow it at all to overwrite the key-based strings, but probably there are use cases where that makes sense.. 🤷‍♀️

We'll see. I hope that it is not used but it allowed to better show the component in lookbook.

toy commented on app/components/enterprise_edition/banner_component.rb:
```diff
@@ -0,0 +1,107 @@
+
+    def description
+      @description || begin
+        I18n.t("ee.upsale.#{feature_key}.description")
+      rescue StandardError
+        I18n.t("ee.upsale.#{feature_key}.description_html")
```
@ulferts Why rescue instead of falling back using default as it is done for title?
